### PR TITLE
Blank Case Search

### PR DESCRIPTION
**Pull request by esoergel**
## Summary

https://dimagi-dev.atlassian.net/browse/USH-1150
https://dimagi-dev.atlassian.net/browse/USH-1105
https://docs.google.com/document/d/1sLHXEFPfj56zcLw8ZZPDXMVOUxqHiIpxKIcp2hE2ueE/edit#
To be deployed with https://github.com/dimagi/formplayer/pull/987

I'm sorry to say it might be easiest to review as a whole - there was a decent amount of churn to make sense of the JS.  I'm still familiarizing myself with this codebase, so feel free to provide opinionated feedback.

See product description below.

## Feature Flag
"Enable synchronous mobile searching and case claiming"

## Product Description
This adds the ability to search for cases missing specific values.  Here's the configuration screen:

image

And here's how it looks in the UI:

image

Values are considered to be blank/empty if they are not set on the case or are set to the empty string.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-3205

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by orangejenny** [comment]
FYI I'm about halfway through reviewing this and should be able to finish it tomorrow.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 1]
```diff
@@ -13,29 +13,41 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
-                return value;
-            }
-            if (model.get("input") === "daterange") {
-                return value.replace("__range__", "").replace("__", separator);
+            if (!_.isString(value)) {
+                return [false, undefined];
             }
+            var values = value.split(selectDelimiter),
+                searchForBlank = _.contains(values, ""),
+                values = _.filter(values, function (val) { return val !== ""; });
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:38]: `_.without(values, "")` would be nice & concise here

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 2]
```diff
@@ -329,15 +316,30 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
-                        $($(".query-input-group")[i]).find('.search-for-blank').prop('checked', includeBlank);
                     }
                 }
                 self.setStickyQueryInputs();
             });
         },
 
+        toggleBlankSearch: function (e) {
+            var self = this,
+                searchForBlank = $(e.target).prop('checked'),
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Nit: I'm a proponent of always using `currentTarget` - doesn't matter here, but `target` can be a descendant of the element that has the event handler.

**Review by orangejenny — COMMENTED** [review]
This looks great.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/details/case_claim.js [hunk 3]
```diff
@@ -259,6 +263,7 @@ hqDefine("app_manager/js/details/case_claim", function () {
 
         if (searchProperties.length > 0) {
             for (var i = 0; i < searchProperties.length; i++) {
+                // searchProperties is a list of CaseSearchProperty objects
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:266]: Oh yeah, that boilerplate was a bit annoying.  I also had trouble figuring out where the raw data from formplayer was initialized into a JS object - would be nice to have that happen in just one place, but it's kinda scattered across a few of these.

File: corehq/apps/case_search/utils.py [hunk 4]
```diff
@@ -119,27 +123,30 @@ def _add_case_property_queries(self):
...
+                value = [re.sub(to_remove, '', val) for val in value]
+            else:
+                value = re.sub(to_remove, '', value)
+        return value
+
+    @cached_property
+    def _patterns_to_remove(self):
+        patterns_by_property = defaultdict(list)
+        for pattern in self.config.ignore_patterns.filter(domain=self.domain, case_type__in=self.case_types):
+            patterns_by_property[pattern.case_property].append(re.escape(pattern.regex))
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/case_search/utils.py:168]: huh, yeah I think you're right - does this mean these patterns don't actually work as regexps?

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html [hunk 5]
```diff
@@ -92,6 +92,8 @@ <h4 class="modal-title">
                           <input type="checkbox" data-bind="checked: isMultiselect"/>
                       </div>
                   </div>
+                {% endif %}
+                {% if js_options.default_value_expression_enabled %}
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html:96]: This is the way the other things under the USH_CASE_CLAIM_UPDATES flag are gated, so I just went along with it.  It also checks a minimum app version, though I guess that's not really relevant for webapps.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js (hunk 2 above)
Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: thanks, I wasn't aware

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 6]
```diff
@@ -358,11 +358,10 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
 
         clearAction: function () {
-            var self = this,
-                fields = $(".query-field");
-            fields.each(function () {
-                this.value = '';
-                $(this).trigger('change.select2');
+            var self = this;
+            _.each(self.collection.models, function (model) {
+                model.set('value', '');
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:356]: yeah, kinda indirectly - the model change triggers a re-render, so `onRender` gets called, which initializes select2.  Now that I'm writing that out, I wonder if I could've used that same logic to call `initMapboxWidget` when needed.

File: corehq/apps/case_search/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/case_search/utils.py:168]: I believe so. The settings page where you configure these expressions literally uses `+`, an invalid regex, as an example.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/case_search/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/case_search/utils.py:205]: That's...unfortunate.  It does look like it only does exact string matches, and that's clear from what's presented to the user, at least.

**Comment by esoergel** [comment]
@orangejenny pinging for a review of the last couple commits - should be tiny, but this now passed QA and is ready for merge

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/es/tests/test_case_search_es.py [hunk 7]
```diff
@@ -557,6 +557,25 @@ def _assert_related_case_ids(self, cases, paths, ids):
...
+        config.fuzzy_properties.add(fuzzy_properties)
+        self._assert_query_runs_correctly(
+            self.domain,
+            cases,
+            CaseSearchCriteria(self.domain, ['song', 'show'], {'description': 'New York'}).search_es,
+            None,
+            ['c1', 'c2']
+        )
+        fuzzy_properties.delete()
+        config.delete()
```
Comment (snopoke) [inline corehq/apps/es/tests/test_case_search_es.py:595]: These should use `self.addCleanup` and be above the assert.

**Review by snopoke — COMMENTED** [review]
Same comment for other tests.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 8]
```diff
@@ -13,87 +13,57 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
-
-        modelEvents: {
-            'change': 'render',
-        },
-
-        geocoderItemCallback: function (addressTopic) {
+        geocoderItemCallback = function (addressTopic, model) {
             return function (item) {
+                model.set('value', item.place_name);
+                initMapboxWidget(model);
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:55]: @esoergel do you recall why it was necessary to call `initMapboxWidget`?

This function handles events from the widget so I don't see how calling `initMapboxWidget` make sense.

Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:55]: I don't recall, but I agree it looks pretty weird.  It was tricky to manage control flow - one guess is maybe something cleared out the input in the widget and re-initializing the widget set it back?

**Pull request merged**